### PR TITLE
#460 HighestOf + unit tests

### PR DESCRIPTION
**Pull request by bedward70**
The initial ticket #460.
* The `Max` class was restored and was rename to `HighestOf`.
* Unit tests were added (the `HighestOfTest` class).

P.S.: The `Max` class was deleted in the https://github.com/yegor256/cactoos/commit/25dc45d Yegor Bugayenko <yegor256@gmail.com> on 11/27/17 at 4:05 AM.

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#575](https://github.com/yegor256/cactoos/issues/575) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#575](https://github.com/yegor256/cactoos/issues/575) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/main/java/org/cactoos/scalar/Folded.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * <p>This class implements {@link Scalar}, which throws a checked
+ * {@link Exception}. This may not be convenient in many cases. To make
+ * it more convenient and get rid of the checked exception you can
+ * use {@link UncheckedScalar} or {@link IoCheckedScalar} decorators.</p>
+ *
+ * @author Alexander Dyadyushenko (gookven@gmail.com)
+ * @author Eduard Balovnev (bedward70@mail.ru)
+ * @version $Id$
+ * @param <T> Scalar type
+ * @since 0.9
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Folded.java]: @bedward70 Let's change to next release (0.29).

File: src/main/java/org/cactoos/scalar/Folded.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ */
+package org.cactoos.scalar;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import org.cactoos.BiFunc;
+import org.cactoos.Scalar;
+
+/**
+ * Folds iterable via BiFunc
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Folded.java]: @bedward70 Can you improve this javadoc? When I should use an object of this class? Can you give us an usage example?

File: src/main/java/org/cactoos/scalar/HighestOf.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.scalar;
+
+import org.cactoos.Scalar;
+import org.cactoos.iterable.IterableOf;
+
+/**
+ * Find the greater among items.
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/HighestOf.java]: @bedward70 Can you improve this javadoc? Can you give us an usage example?

File: src/main/java/org/cactoos/scalar/HighestOf.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * <p>There is no thread-safety guarantee.
+ *
+ * @author Fabricio Cabral (fabriciofx@gmail.com)
+ * @author Eduard Balovnev (bedward70@mail.ru)
+ * @version $Id$
+ * @param <T> Scalar type
+ * @see UncheckedScalar
+ * @see IoCheckedScalar
+ * @since 0.10
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/HighestOf.java]: @bedward70 Let's change to next release (0.29).

File: src/main/java/org/cactoos/scalar/HighestOf.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+ */
+public final class HighestOf<T extends Comparable<T>> implements Scalar<T> {
+
+    /**
+     * Items.
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/HighestOf.java]: @bedward70 Items or just the highest value?

File: src/test/java/org/cactoos/scalar/FoldedTest.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link Folded}.
+ *
+ * @author Eduard Balovnev (bedward70@mail.ru)
+ * @version $Id$
+ * @since 0.10
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/FoldedTest.java]: @bedward70 Let's change to next release (0.29).

File: src/test/java/org/cactoos/scalar/HighestOfTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link HighestOf}.
+ *
+ * @author Fabricio Cabral (fabriciofx@gmail.com)
+ * @author Eduard Balovnev (bedward70@mail.ru)
+ * @version $Id$
+ * @since 0.10
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HighestOfTest.java]: @bedward70 Let's change to next release (0.29).

File: src/test/java/org/cactoos/scalar/FoldedTest.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+ */
+public final class FoldedTest {
+
+    @Test(expected = NoSuchElementException.class)
+    public void emptyListTest() throws Exception {
+        new Folded<>(
+            (first, last) -> first,
+            Collections.emptyList()
+        )
+            .value();
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/FoldedTest.java]: @bedward70 Please, could you fix the indentation? `).value()`. I don't know why Qulice doesn't catch this type of error.

File: src/test/java/org/cactoos/scalar/FoldedTest.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    @Test
+    public void singleTest() throws Exception {
+        final Integer single = 10;
+        MatcherAssert.assertThat(
+            "Can't find the single",
+            new Folded<>(
+                (first, last) -> first,
+                new IterableOf<Scalar<Integer>>((Scalar<Integer>) () -> single)
+            )
+                .value(),
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/FoldedTest.java]: @bedward70 Please, could you fix the indentation? `).value()`.

File: src/test/java/org/cactoos/scalar/FoldedTest.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            "Can't find the first",
+            new Folded<>(
+                (first, last) -> first,
+                new IterableOf<Scalar<String>>(
+                    (Scalar<String>) () -> one,
+                    (Scalar<String>) () -> two,
+                    (Scalar<String>) () -> three
+                )
+            )
+                .value(),
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/FoldedTest.java]: @bedward70 Please, could you fix the indentation? `).value()`.

File: src/test/java/org/cactoos/scalar/FoldedTest.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+            "Can't find the last",
+            new Folded<>(
+                (first, last) -> last,
+                new IterableOf<Scalar<Character>>(
+                    (Scalar<Character>) () -> one,
+                    (Scalar<Character>) () -> two,
+                    (Scalar<Character>) () -> three
+                )
+            )
+                .value(),
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/FoldedTest.java]: @bedward70 Please, could you fix the indentation? `).value()`.

File: src/test/java/org/cactoos/scalar/HighestOfTest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class HighestOfTest {
+
+    @Test(expected = NoSuchElementException.class)
+    public void maxAmongEmptyTest() throws Exception {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HighestOfTest.java]: @bedward70 Change `max` to `highest`.

File: src/test/java/org/cactoos/scalar/HighestOfTest.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        new HighestOf<>(() -> Collections.emptyIterator()).value();
+    }
+
+    @Test
+    public void maxAmongOneTest() throws Exception {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HighestOfTest.java]: @bedward70 Change `max` to `highest`.

File: src/test/java/org/cactoos/scalar/HighestOfTest.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        final int num = 10;
+        MatcherAssert.assertThat(
+            "Can't find the greater among one",
+            new HighestOf<Integer>(() -> new Integer(num)).value(),
+            Matchers.equalTo(num)
+        );
+    }
+
+    @Test
+    public void maxAmongManyTest() throws Exception {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HighestOfTest.java]: @bedward70 Change `max` to `highest`.

File: src/test/java/org/cactoos/scalar/HighestOfTest.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    @Test(expected = NoSuchElementException.class)
+    public void maxAmongEmptyTest() throws Exception {
+        new HighestOf<>(() -> Collections.emptyIterator()).value();
+    }
+
+    @Test
+    public void maxAmongOneTest() throws Exception {
+        final int num = 10;
+        MatcherAssert.assertThat(
+            "Can't find the greater among one",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HighestOfTest.java]: @bedward70 Change `greater` to `highest`.

File: src/test/java/org/cactoos/scalar/HighestOfTest.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+            new HighestOf<Integer>(() -> new Integer(num)).value(),
+            Matchers.equalTo(num)
+        );
+    }
+
+    @Test
+    public void maxAmongManyTest() throws Exception {
+        final int num = 10;
+        MatcherAssert.assertThat(
+            "Can't find the greater among many",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HighestOfTest.java]: @bedward70 Change `greater` to `highest`.

File: src/test/java/org/cactoos/scalar/HighestOfTest.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * Test case for {@link HighestOf}.
+ *
+ * @author Fabricio Cabral (fabriciofx@gmail.com)
+ * @author Eduard Balovnev (bedward70@mail.ru)
+ * @version $Id$
+ * @since 0.10
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class HighestOfTest {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HighestOfTest.java]: @bedward70 Could you add more tests? Use `Double`, negative numbers, `Double.MAX_VALUE`, `Double.MIN_VALUE`, `Integer.MAX_VALUE`, `Integer.MIN_VALUE`, `NaN` (Not a Number), `DateOf`, `SumOf`, etc.

File: src/main/java/org/cactoos/scalar/HighestOf.java [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    private final Scalar<T> result;
+
+    /**
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/HighestOf.java]: @bedward70 Please, could you add two more ctor: `public HighestOf(final T... items)` and `public HighestOf(final Iterable<T>... items)`

File: src/main/java/org/cactoos/scalar/HighestOf.java [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * @author Fabricio Cabral (fabriciofx@gmail.com)
+ * @author Eduard Balovnev (bedward70@mail.ru)
+ * @version $Id$
+ * @param <T> Scalar type
+ * @see UncheckedScalar
+ * @see IoCheckedScalar
+ * @since 0.10
+ */
+public final class HighestOf<T extends Comparable<T>> implements Scalar<T> {
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/HighestOf.java]: @bedward70 You have said: *I brought the Max class back, but I rename it to HighestOf*, but `MaxOf` still there here. Is it correct?

Comment (bedward70) [reply src/main/java/org/cactoos/scalar/HighestOf.java]: @fabriciofx  Yes, it is correct, but `MaxOf` and `MinOf` are intended for `Number` implementations only.
The`HighestOf` and `LowestOf` are intended for `Comparable` implementations.
See details: #460 and #475.

File: src/main/java/org/cactoos/scalar/HighestOf.java (hunk 4 above)
Comment (bedward70) [reply src/main/java/org/cactoos/scalar/HighestOf.java]: @fabriciofx  "llorllale please change to 1.0". See: 
https://github.com/yegor256/cactoos/pull/477/files/86156dec198aee0b1697cb26afec58446c2a6115#diff-6c2da6f775cf87e8034fa7f06f25211a

File: src/test/java/org/cactoos/scalar/FoldedTest.java (hunk 6 above)
Comment (bedward70) [reply src/test/java/org/cactoos/scalar/FoldedTest.java]: @fabriciofx  "llorllale please change to 1.0". See: 
https://github.com/yegor256/cactoos/pull/477/files/86156dec198aee0b1697cb26afec58446c2a6115#diff-6c2da6f775cf87e8034fa7f06f25211a

File: src/main/java/org/cactoos/scalar/HighestOf.java (hunk 18 above)
Comment (bedward70) [reply src/main/java/org/cactoos/scalar/HighestOf.java]: @fabriciofx `public HighestOf(final T... items)` has done.
I can't create `public HighestOf(final Iterable<T>... items)` ctor:  
The `public HighestOf(final Iterable<T>... items)` and the `public HighestOf(final Iterable<Scalar<T>> iterable)` have the same JAVA semantics.

**Comment by bedward70** [comment]
@fabriciofx I tried to fix all findings. Please, review.
Please specify the release number: 0.29 or 1.0.
Thank you.

**Comment by fabriciofx** [comment]
@bedward70 Well, I'm always using the next release version and Yegor doesn't complain about it. So, use the 0.29 version.

**Comment by bedward70** [comment]
@fabriciofx I regrouped tests in the `HighestOfTest` and set the 0.29 version.

**Comment by 0crat** [comment]
@fabriciofx[/z](https://www.0crat.com/u/fabriciofx) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0crat** [comment]
The user @fabriciofx[/z](https://www.0crat.com/u/fabriciofx) resigned from [#575](https://github.com/yegor256/cactoos/issues/575), please stop working

**Comment by 0crat** [comment]
This pull request [#575](https://github.com/yegor256/cactoos/issues/575) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C63314D6Z/432133017). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by amihaiemil** [comment]
@bedward70 ping

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by bedward70** [comment]
@amihaiemil I reviewed the JavaDoc of `HighestOf` and changed the title and description

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by amihaiemil** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@rultor merge

**Comment by amihaiemil** [comment]
@bedward70 there seems to be some conflict, can you look into it?

**Comment by bedward70** [comment]
@amihaiemil I resolved the conflict. The PR still contains two files: `HighestOf` and `HighestOfTest`.

**Comment by amihaiemil** [comment]
@yegor256 can we try to merge again?

**Comment by amihaiemil** [comment]
@bedward70 Don't forget to mention the initial ticket in the PR's description, please

**Comment by bedward70** [comment]
@amihaiemil I was added #460 to description.

**Comment by amihaiemil** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job, as in [§30](http://datum.zerocracy.com/pages/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
QA review completed: +15 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka), total is [+480](https://www.0crat.com/u/ypshenychka)

**Comment by 0crat** [comment]
Order was finished, quality was "good": +20 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+870](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+10589](https://www.0crat.com/u/yegor256)
